test/extended/util/disruption/disruption.go

```diff
@@ -197,71 +221,76 @@ func (cma *chaosMonkeyAdapter) Test(sem *chaosmonkey.Semaphore) {
 	cma.test.Test(cma.framework, sem.StopCh, cma.UpgradeType)
 }
 
-func finalizeTest(start time.Time, tc *junit.TestCase, ts *junit.TestSuite, f *framework.Framework) {
+func finalizeTest(start time.Time, testName, className string, ts *junitapi.JUnitTestSuite, f *framework.Framework) {
 	now := time.Now().UTC()
-	tc.Time = now.Sub(start).Seconds()
+	testDuration := now.Sub(start).Seconds()
+
+	// r is the primary means we are informed of test results here. We expect ginko to panic on any failure so
+	// if r is nil, we passed. If not, we start checking if this was a flake or fail below.
 	r := recover()
 
 	// if the framework contains additional test results, add them to the parent suite or write them to disk
 	for _, summary := range f.TestSummaries {
+
 		if test, ok := summary.(additionalTest); ok {
-			testCase := &junit.TestCase{
-				Name: test.Name,
-				Time: test.Duration.Seconds(),
+			tc := &junitapi.JUnitTestCase{
+				Name:      test.Name,
+				Classname: className,
+				Duration:  test.Duration.Seconds(),
 			}
 			if len(test.Failure) > 0 {
-				testCase.Failures = append(testCase.Failures, &junit.Failure{
-					Message: test.Failure,
-					Value:   test.Failure,
-				})
+				tc.FailureOutput = &junitapi.FailureOutput{Message: test.Failure}
 			}
-			ts.TestCases = append(ts.TestCases, testCase)
+			ts.TestCases = append(ts.TestCases, tc)
 			continue
 		}
 
-		filePath := filepath.Join(framework.TestContext.ReportDir, fmt.Sprintf("%s_%s_%s.json", summary.SummaryKind(), filesystemSafeName(tc.Name), now.Format(time.RFC3339)))
+		filePath := filepath.Join(framework.TestContext.ReportDir, fmt.Sprintf("%s_%s_%s.json", summary.SummaryKind(), filesystemSafeName(testName), now.Format(time.RFC3339)))
 		if err := ioutil.WriteFile(filePath, []byte(summary.PrintJSON()), 0644); err != nil {
 			fmt.Fprintf(os.Stderr, "error: Failed to write file %v with test data: %v\n", filePath, err)
 		}
 	}
 
 	if r == nil {
+		// Test can be considered successful, but may have flaked, see below:
+		ts.TestCases = append(ts.TestCases, &junitapi.JUnitTestCase{
+			Name:      testName,
+			Classname: className,
+			Duration:  testDuration,
+		})
```

Comment:
Because we no longer pre-create TestCase objects before the test starts, this is the place where successful tests are added.